jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/AttachHandler.java

```diff
@@ -522,6 +615,51 @@ static final class teardownHook extends Thread {
 			super("Attach API teardown"); //$NON-NLS-1$)
 		}
 
+		private void threadJoinHelper(Thread thread, long shutdownDeadlineNs) {
+			long timeout = 100;
+			int retryNumber = 0;
+			for (;;) {
+				long currentNanoTime = System.nanoTime();
+				if (currentNanoTime >= shutdownDeadlineNs) {
+					// already reached shutdownDeadlineNs
+					IPC.logMessage(thread + " already reached shutdownDeadlineNs : currentNanoTime = " + currentNanoTime ); //$NON-NLS-1$
+					break;
+				}
+				long tempTimeout = NANOSECONDS.toMillis(shutdownDeadlineNs - currentNanoTime);
+				if (tempTimeout < timeout) {
+					// not wait beyond shutdownDeadlineNs
+					timeout = tempTimeout;
```

Comment:
`timeout` could become zero here: was that intended?
Should it be reset to 100 on each iteration of the loop?

Reply:
No, added `break;` when `tempTimeout` is `0`.
`timeout` is doubled after each `join()` timed out however it can't be greater than the remaining shutdown timeout, i.e., the `tempTimeout`.

Reply:
I think reversing the test is easier to read:
```java
				if (timeout > tempTimeout) {
					// not wait beyond shutdownDeadlineNs
					timeout = tempTimeout;
				}
```

Reply:
Reverted